runtime/vm/threadhelp.cpp

```diff
@@ -223,58 +232,105 @@ monitorWaitImpl(J9VMThread *vmThread, j9object_t object, I_64 millis, I_32 nanos
  * @return 0 on success, non-zero on failure. This function always sets the current exception on failure.
  */
 IDATA
-threadSleepImpl(J9VMThread* vmThread, I_64 millis, I_32 nanos)
+threadSleepImpl(J9VMThread *vmThread, I_64 millis, I_32 nanos)
 {
-	IDATA rc;
-	J9JavaVM* javaVM = vmThread->javaVM;
-
-//	Trc_JCL_sleep_Entry(vmThread, millis, nanos);
+	IDATA rc = 0;
+	J9JavaVM *javaVM = vmThread->javaVM;
 
-	if (validateTimeouts(vmThread, millis, (I_32)nanos)) {
-//		Trc_JCL_sleep_Exit(vmThread);
-		return -1;
-	}
-	
+	/* Trc_JCL_sleep_Entry(vmThread, millis, nanos); */
+	if (0 == validateTimeouts(vmThread, millis, nanos)) {
 #ifdef J9VM_OPT_SIDECAR
-	/* Increment the wait count even if the deadline is past */
-	vmThread->mgmtWaitedCount++;
+		/* Increment the wait count even if the deadline is past. */
+		vmThread->mgmtWaitedCount++;
 #endif
 
-	TRIGGER_J9HOOK_VM_SLEEP(javaVM->hookInterface, vmThread, millis, nanos);
-	internalReleaseVMAccessSetStatus(vmThread, J9_PUBLIC_FLAGS_THREAD_SLEEPING);
-	rc = omrthread_sleep_interruptable(millis, nanos);
-	internalAcquireVMAccessClearStatus(vmThread, J9_PUBLIC_FLAGS_THREAD_SLEEPING);
-	TRIGGER_J9HOOK_VM_SLEPT(javaVM->hookInterface, vmThread);
+		TRIGGER_J9HOOK_VM_SLEEP(javaVM->hookInterface, vmThread, millis, nanos);
+		internalReleaseVMAccessSetStatus(vmThread, J9_PUBLIC_FLAGS_THREAD_SLEEPING);
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+		/* Thread.sleep() adjustment only supports CRIURestoreNonPortableMode which is default mode. */
+		BOOLEAN adjustmentRequired = 0 == javaVM->checkpointState.checkpointRestoreTimeDelta;
+		I_64 beforeSleep = 0;
+		if (adjustmentRequired) {
+			PORT_ACCESS_FROM_JAVAVM(javaVM);
+			beforeSleep = j9time_nano_time();
+		}
+#endif /* defined(J9VM_OPT_CRIU_SUPPORT) */
+		rc = omrthread_sleep_interruptable(millis, nanos);
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+		if (adjustmentRequired && (0 != javaVM->checkpointState.checkpointRestoreTimeDelta)) {
+			/* There was a checkpoint. */
+			PORT_ACCESS_FROM_JAVAVM(javaVM);
+			const I_32 oneMillion = 1000000;
+			I_64 sleptTimeNanos = j9time_nano_time() - beforeSleep;
+			I_64 sleptTimeMillis = sleptTimeNanos / oneMillion;
+			BOOLEAN continueSleepRequired = TRUE;
+
+			if (sleptTimeMillis < millis) {
```

Comment:
sleep can also be interrupted

Reply:
I did consider the case it is interrupted, set `continueSleepRequired` to `true`, and continue `omrthread_sleep_interruptable()` afterwards.
Just realize that the second one could be interrupted as well.